### PR TITLE
Fix ext_proc filter can not send non-utf8 character by gRPC

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
Currently ext_proc is using protobuf string type to transmit header and trailer value. However, protobuf string requires it to be valid UTF-8 character.  This has caused issues so a temporary workaround is applied to ext_proc filter by converting those non utf-8 character header value into "!" before sending it to the ext_proc server.
However, this approach brings in new issue that the header or trailer it sends to ext_proc server is completely changed.

To solve this problem a new API , like HeaderMpaBytes, is needed that uses the bytes type instead of string. bytes does not need to be UTF-8.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yanjunxiang-google** [comment]
@yanavlasov @htuch @tyxia @mattklein123

**Comment by yanjunxiang-google** [comment]
For reviewers, converting non-utf8 character to "!" when ext_proc filter sending header/trailers to ext_proc server is problematic.  It completely changed the original header or trailer value.

This PR take the advantage that the ext_proc filter is still in Alpha state, to resolve the issue completely.

File: api/envoy/config/core/v3/base.proto
Comment (htuch) [inline api/envoy/config/core/v3/base.proto:346]: Did you think of adding a `value_bytes` field to the existing `HeaderValue` and treating it as a one-off. The advantage is that we could start to upgrade existing APIs in-place. For ext_proc situation, I think it might be OK to make this change given the WiP statement, although @mattklein123 has some concerns that these APIs are now in heavy use.

Comment (yanavlasov) [reply api/envoy/config/core/v3/base.proto:346]: Upgrade of APIs in place can be problematic. How would a client know if a server supports reading from both fields?

Comment (mattklein123) [reply api/envoy/config/core/v3/base.proto:346]: If the entire API is still marked alpha/WiP, I guess it's OK, but I do get the impression that people are using this thing. I would have this be the last change and then mark it non-alpha, or just mark it non-alpha now and start with the standard API change semantics that we do for other stable APIs.

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
Waiting for the API approval.

File: api/envoy/config/core/v3/base.proto
Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto:346]: Thanks for the comments!

I just added the corresponding change in the ProcessingResponse proto message to make non-utf8 character supported in both directions (Envoy <---> ext_proc server) .

Regarding should we add a separate message,  or just add a one-of field, I think one-of is a good idea.  It avoids defining largely duplicated messages, like HeaderValueOption ---> HeaderValueOptionBytes.  But I don't fully understand why it has the advantage of upgrading APIs in place. Could you please elaborate @htuch ?  IIUC, either approach is a breaking change, which needs ext_proc filter and server upgraded at same time with a little bit downtime.

cc @adisuissa for his opinion.

Comment (htuch) [reply api/envoy/config/core/v3/base.proto:346]: You could populate both fields for a period of time, under runtime control, to avoid ext_proc / server needing to upgrade at same time.

File: api/envoy/config/core/v3/base.proto
Comment (tyxia) [inline api/envoy/config/core/v3/base.proto:334]: This option is redundant? As you are adding a new message here

File: api/envoy/config/core/v3/base.proto
Comment (tyxia) [inline api/envoy/config/core/v3/base.proto:431]: We don't need deprecated field in new proto message I think

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (tyxia) [inline api/envoy/service/ext_proc/v3/external_processor.proto:198]: So I am wondering have we decided how to deal with breaking API change, if they are many clients out there?

At least I think we need release note here.

File: api/envoy/config/core/v3/base.proto
Comment (tyxia) [inline api/envoy/config/core/v3/base.proto:399]: ditto. this is new message.  And please check other places in this PR.

**Review by tyxia — CHANGES_REQUESTED** [review]
Thanks for working on this!

File: api/envoy/config/core/v3/base.proto
Comment (htuch) [reply api/envoy/config/core/v3/base.proto:346]: Right. Yeah, it would be useful to get @yanavlasov and @envoyproxy/api-shepherds take on this, it's a pretty fundamental change if we decide to extend `HeaderValue` this way (with `value_bytes`). The tradeoffs I see is it allows us to retrofit proper header value encoding for UTF-8. The downside is that it's a shared proto used in many places and it wil lbe hard to tell which ones are using the bytes vs. string representation.

Comment (adisuissa) [reply api/envoy/config/core/v3/base.proto:346]: > You could populate both fields for a period of time, under runtime control, to avoid ext_proc / server needing to upgrade at same time.

This is a de-facto a deprecation and removal of the old field. The alternative of populating both fields is worse IMHO (large message-sizes).

> I guess adding runtime flag applies to other filters, like ext_authz, if they plan to move to the value_bytes in case they want to fix the non-utf8 character issue. For the ext_proc filter code here, we can directly populate the value_bytes since it is in Alpha state, right?

I'm not sure it is possible, for example if the ext_proc server still uses the old proto definition, i won't be able access these fields.

An alternative could be to add a some encoding/decoding ability to send the value as a string. This can be accompanied with some field that signals when the sent string is encoded.

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto:346]: Thanks @htuch . The downside you mentioned does raise a concern to me. In that regards, using a separate message can make the change impact localize to the ext_proc  itself, which probably is better solution for the short-term goal?.

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto:346]: @adisuissa  the alternative you mentioned by encoding a non-utf8 character into string before sending it is still a breaking change with some complexity.  If we can make breaking change, send it as bytes seem to be simpler.

File: api/envoy/config/core/v3/base.proto
Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto]: We are using oneof approach in the HeaderValue message. That approach make it possible to make this PR a non-breaking change.  With that, the new sets of messages are removed.

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto]: Yes, a runtime flag is added to protect the original behavior.  However, it might be ideal set the run-time flag to a value to make breaking change as default to get the ext_proc filter to the state which by default support non-utf8 characters.  Let's see what other people think.

File: api/envoy/config/core/v3/base.proto
Comment (yanavlasov) [inline api/envoy/config/core/v3/base.proto:327]: The HeaderValue message is used in other protobufs that are marked `stable`. Changing to oneof is a breaking change and we can not do it this way.

I think you need to drop oneof and use annotation: [(udpa.annotations.field_migrate).oneof_promotion = "value"];

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/config/core/v3/base.proto
Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto]: I changed the code to have ext_proc by default support non-utf8 character encoding since IMHO this is a bug fix. For customers who want to maintain the old behavior, they can toggle the run time flag to be false.

File: api/envoy/config/core/v3/base.proto
Comment (tyxia) [reply api/envoy/config/core/v3/base.proto:327]: +1.  Besides, I think you also then should add config validation to enforce only one field is set.   

Basically:
1)  Add `[(udpa.annotations.field_migrate).oneof_promotion = "YOUR_NEW_NAME_FOR_ONEOF"` to the existing `value` field. Also, add the same annotation to the `value_bytes` field.
2) Add the comments saying only one of two is specified 
3) Add config validation to enforce this, throw exception otherwise.

BTW, [Here](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#field-level) explains why in details.

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto:327]: I see. More background about this: https://github.com/aip-dev/google.aip.dev/issues/393

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto]: Thanks for the detail comments!
There is also an example here:  https://github.com/envoyproxy/envoy/blob/35514452d10ac56f5afca0c177149f0cdd030aa7/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L783.

File: api/envoy/config/core/v3/base.proto
Comment (tyxia) [reply api/envoy/config/core/v3/base.proto:334]: yea, as I mentioned what you need is adding `one_of_promotion` to this new field as well.
```
bytes value_bytes = 3 [(validate.rules).bytes = {min_len: 0 max_len: 16384}];
(udpa.annotations.field_migrate).oneof_promotion = "value_type"
```
Basically this annotations is a reminder to future major API update that coverts those field to to oneof. This is the reason why I also suggested  the configuration validation . Because in current API, those two fields are technically not oneof. They will become oneof in next API promotion (if happens)  

I think that http_connection_manager.proto didn't do it right.

Comment (tyxia) [reply api/envoy/config/core/v3/base.proto:334]: [Here](https://github.com/envoyproxy/envoy/blame/61958d40059a3d901b6f0e453d769f58a3dd2e6a/api/envoy/config/route/v3/route_components.proto#LL796C12-L796C26) is a better example: `cluster_header` and `cluster`

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto:334]: How about the stats_flush in bootstrap.proto?https://github.com/envoyproxy/envoy/blob/61958d40059a3d901b6f0e453d769f58a3dd2e6a/api/envoy/config/bootstrap/v3/bootstrap.proto#L198

BTW, that cluster_header is also promoted to a oneof type: cluster_specifier here: https://github.com/envoyproxy/envoy/blob/61958d40059a3d901b6f0e453d769f58a3dd2e6a/api/envoy/config/route/v3/route_components.proto#L1019

Comment (tyxia) [reply api/envoy/config/core/v3/base.proto:334]: That `cluster_specifier` you referring to is in `RouteAction`. which is not related to the example I shared. In other words, It is not the promoted result of my example. You can also see it from the timeline that it was added 4 years ago while my example is 1 year ago.

IMHO oneof you have here is redundant. It is oneof with only single element.

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto:334]: @adisuissa could you please comment on this?  My intuitively interprectation of the text below from the [Style Guide](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#field-level) is that the new field is a oneof field, and the existing field will be moved into that oneof in the next major API change.
"
[(udpa.annotations.field_migrate).oneof_promotion = "<oneof name>"] to denote that the field will be promoted to a given oneof in the next API major version.
"

Comment (htuch) [reply api/envoy/config/core/v3/base.proto:334]: There is not going to be another major xDS API major version for a long time (if ever). In the meanwhile, we use the `oneof_promotion` fields to indicate that this is logically a `oneof`, even though it's not enforced in the proto type system. The advice in https://github.com/envoyproxy/envoy/pull/27865/files#r1232707830 will work.

Comment (yanjunxiang-google) [reply api/envoy/config/core/v3/base.proto]: I see. Thanks for the explanation.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by yanavlasov** [comment]
LGTM. Please merge main.

/wait

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: Is that possible `sh.has_header ` (i.e. we have key field) but  both `value() and value_bytes()` are empty? It seems so as there is no min_len 1 rule for value fields. 
Is this situation handled by code or can we do early return here?

We can follow up on this in future PRs (if needed)

**Review by tyxia — APPROVED** [review]
Thank you!

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: The header value content check is done the logic below: Http::HeaderUtility::headerValueIsValid(header_value)).
From this logic and the HeaderValue proto definition(no min length for value field), it looks to me empty value content is valid, and will pass the follow up mutation rule check by default.

That's the existing logic, and this PR does not change that, i.e., we do not bother if both fields are empty.

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: Yes, I think it will pass the headerValueIsValid check, but do we need to go through the whole check here if it is already empty. This is what I meant by early return.

I understand it is existing logic so I suggested we can follow up on this if needed.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: An empty header value still needs to go through the mutation rule check. For example, if disallow_all is set to true, that header contains empty header value will still be rejected:

https://github.com/envoyproxy/envoy/blob/f567f66fa68b1533e9daffbdd5c5fd31ee65cba5/source/extensions/filters/common/mutation_rules/mutation_rules.cc#L72

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: But what mutation will happen if sh.header()'s value fields are empty in the first place. That essentially is same as sh.has_header() = false, which skips all the processing here. no?

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: no. sh.has_header() checks whether the header key/value pair exists.

In that pair, the key is more important than value. Like all the system headers check, allow_expression, disallow_expression check are all based on key.

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: I meant what kinds of mutation will happen here in `applyHeaderMutations` if `mutation.set_headers` only has key but not header. 
In other words, what is the mutation outcome not what does check do. Check is the step before mutation. but if there is no outcome at all, does check still important?

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: Yeah, if the ext_proc server sends back a set_header as key/value:   "foo":  "",  this header will be added to the original header list.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: But this discussion raise a good question that what if the "key" in the header is empty.  

It looks to me an empty header_key will return true in the headerNameIsValie() check in this utility function:
https://github.com/envoyproxy/envoy/blob/3a68249d6b5039f7df5c3f7d740af20a89d97b53/source/common/http/header_utility.cc#L204

Is this expected @yanavlasov ?

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: Hmm, but I am wondering what will such empty value header`"foo": ""` do in practice ?  Probably it is allowed in RFC but don't make too much difference in practice.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: ext_proc filter is relying on these utilities to make sure the key/value pair are spec compliant.  Adding separate checks in ext_proc filter does not feel right to me.

In either case, let's deal with it separately since it's an existing issue and not related to this adding value_bytes() to support non-utf8 characters.

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:86]: Sure SGTM, let's discuss and address it in next PR (which is also the original intention of my comment)

**Comment by yanjunxiang-google** [comment]
Kind Ping!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
